### PR TITLE
Mobx Legend improvements for TableMixin

### PR DESCRIPTION
**Pull request by nf-s**
### What this PR does

Fixes https://github.com/TerriaJS/terriajs/issues/4245

Related to https://github.com/TerriaJS/terriajs/pull/4120

* Add `multipleTitles` and `maxMultipleTitlesShowed` to `LegendItemTraits`
* Aggregate legend items in `createLegendItemsFromEnumColorMap` by colour, that is merge legend items with the same colour (using `multipleTitles`)
* Only generate `tableStyles` for region columns if no other styles exist

merged legend items

### Examples

Not generating region-column styles:  http://ci.terria.io/mobx-table-style-mods/#share=s-njw3Wy76cGdMXlpKDeNgfMMlWCh

Add `NSW LGA NEXIS 2012` layer and `POPULATION ESTIMATE` should be default:   http://ci.terria.io/mobx-table-style-mods/

Generating region-column styles: http://ci.terria.io/mobx-table-style-mods/#share=s-qtpyWzyZjPxyPAFd7jnGBBklMyJ

### Checklist

-   [x] Tests for this are handled by https://github.com/TerriaJS/terriajs/pull/4219 - see https://github.com/TerriaJS/terriajs/pull/4219/files#diff-76cab554c85da4157534e0a1c0e24386
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Comment by nf-s** [comment]
what do you think @rowanwins @steve9164 ?

File: lib/ModelMixins/TableMixin.ts [hunk 1]
```diff
@@ -102,7 +102,20 @@ export default function TableMixin<T extends Constructor<Model<TableTraits>>>(
       const value = super.activeStyle;
       if (value !== undefined) {
         return value;
-      } else if (this.styles && this.styles.length > 0) {
+      } else if (
+        this.styles &&
+        this.styles.length > 0 &&
+        isDefined(this.styles[0].id)
```
Comment (tephenavies) [inline lib/ModelMixins/TableMixin.ts]: Why is `isDefined(this.styles[0].id)` needed?

**Review by tephenavies — CHANGES_REQUESTED** [review]
I think we should discuss the behaviour and come up with better default style selection before we merge this.

I don't think we should default to a style based on a region column when you drag a region-mapped csv onto the map, but maybe there's another way to address this (e.g. maybe we shouldn't generate a single-colour colour scheme unless there are no other styles available).

File: lib/ModelMixins/TableMixin.ts (hunk 1 above)
Comment (nf-s) [reply lib/ModelMixins/TableMixin.ts:108]: Just to get rid of TypeError for `this.tableColumns.find(col => col.name === this.styles[i].id)`

File: lib/ReactViews/Workbench/Controls/Legend.jsx [hunk 2]
```diff
@@ -207,7 +207,24 @@ const Legend = createReactClass({
...
+            <div
+              title={
+                isDefined(legendItem.multipleTitles)
+                  ? legendItem.multipleTitles.join(", ")
+                  : legendItem.title
+              }
+            >
+              {isDefined(legendItem.multipleTitles)
+                ? `${[...legendItem.multipleTitles]
+                    .splice(0, legendItem.maxMultipleTitlesShowed)
```
[the remark this answers is not in the document]
Comment (nf-s) [reply lib/ReactViews/Workbench/Controls/Legend.jsx]: ah yep! that makes much more sense

File: lib/ModelMixins/TableMixin.ts (hunk 1 above)
Comment (nf-s) [reply lib/ModelMixins/TableMixin.ts]: I can't remember, It must be residual from an earlier version. All of this is pretty rough.

**Comment by nf-s** [comment]
I totally agree, I think only generating the single-colour scheme if nothing else is available would be definitely better than my approach

**Comment by nf-s** [comment]
I have made the requested changes, let me know if you want me to look into what you suggested, or if you have any other ideas!
I'm not that familiar with `Table` stuff in Terria.

**Comment by nf-s** [comment]
> I think we should discuss the behaviour and come up with better default style selection before we merge this.
> 
> I don't think we should default to a style based on a region column when you drag a region-mapped csv onto the map, but maybe there's another way to address this (e.g. maybe we shouldn't generate a single-colour colour scheme unless there are no other styles available).

I have addressed this by only generate tableStyles for region columns if no other styles exist

File: lib/Table/TableAutomaticStylesStratum.ts [hunk 3]
```diff
@@ -258,11 +260,24 @@ export class ColorStyleLegend extends LoadableStratum(LegendTraits) {
...
+      }
+      return prev;
+    }, {});
+
+    return Object.entries(colorMapValues)
+      .map(value => {
         return createStratumInstance(LegendItemTraits, {
-          title: value,
-          color: colorMap.colors[i].toCssColorString()
+          multipleTitles: value[1],
```
Comment (KeyboardSounds) [inline lib/Table/TableAutomaticStylesStratum.ts]: Could you put a comment in here explaining why you're using these particular indices to the value array?

**Review by KeyboardSounds — CHANGES_REQUESTED** [review]

File: lib/Table/TableAutomaticStylesStratum.ts (hunk 3 above)
Comment (nf-s) [reply lib/Table/TableAutomaticStylesStratum.ts:279]: It is from `Object.entries(colorMapValues)`, so `value` is `[key, value]`

Comment (nf-s) [reply lib/Table/TableAutomaticStylesStratum.ts]: It is from `Object.entries(colorMapValues)`, so `value` is `[key, value]`

Comment (KeyboardSounds) [reply lib/Table/TableAutomaticStylesStratum.ts]: OH duh sorry of course

**Review by KeyboardSounds — APPROVED** [review]

File: lib/Traits/CatalogMemberTraits.ts [hunk 4]
```diff
@@ -39,6 +39,13 @@ export class ShortReportTraits extends ModelTraits {
     description: "The content of the section."
   })
   content?: string;
+
```
Comment (nf-s) [inline lib/Traits/CatalogMemberTraits.ts]: This is already in `next` not sure why it is showing a diff

Comment (KeyboardSounds) [reply lib/Traits/CatalogMemberTraits.ts]: merge in `next` and see if it disappears?

File: lib/Table/TableAutomaticStylesStratum.ts [hunk 5]
```diff
@@ -93,8 +92,9 @@ export default class TableAutomaticStylesStratum extends LoadableStratum(
   get defaultChartStyle(): StratumFromTraits<TableStyleTraits> | undefined {
     const scalarColumns = this.catalogItem.tableColumns.filter(
       column =>
-        column.type === TableColumnType.scalar ||
-        column.type === TableColumnType.time
+        !this.catalogItem.excludeStyles?.includes(column.name) &&
```
Comment (tephenavies) [inline lib/Table/TableAutomaticStylesStratum.ts]: Why is it necessary to add `excludeStyles`? We already have `"type": "hidden"`. Are there situations where you want a column to be a special type (it looks like `excludeStyles` is only relevant if a column is "scalar" or "time") and not have a style for that column?

Comment (nf-s) [reply lib/Table/TableAutomaticStylesStratum.ts]: You're right, absolutely not necessary - i have removed it

File: lib/Traits/TableTraits.ts [hunk 6]
```diff
@@ -9,6 +9,7 @@ import primitiveTrait from "./primitiveTrait";
 import RasterLayerTraits from "./RasterLayerTraits";
 import TimeVaryingTraits from "./TimeVaryingTraits";
 import LayerOrderingTraits from "./LayerOrderingTraits";
+import primitiveArrayTrait from "./primitiveArrayTrait";
```
Comment (tephenavies) [inline lib/Traits/TableTraits.ts]: Unnecessary import

**Review by tephenavies — APPROVED** [review]

**Pull request merged**